File: scase-core/src/main/scala/io/jobial/scase/core/Service.scala
package io.jobial.scase.core

trait Service[F[_]] {

  def start: F[ServiceState[F]]

  def startAndJoin: F[ServiceState[F]]
}

trait ServiceState[F[_]] {

  def stop: F[ServiceState[F]]

  def join: F[ServiceState[F]]

  def service: Service[F]
}

File: scase-core/src/main/java/io/jobial/scase/core/javadsl/Service.java
package io.jobial.scase.core.javadsl;

import cats.effect.IO;

import java.util.concurrent.CompletableFuture;

import static io.jobial.scase.core.javadsl.JavaUtils.ioToCompletableFuture;

public class Service {

    private io.jobial.scase.core.Service<IO> service;

    public Service(io.jobial.scase.core.Service<IO> service) {
        this.service = service;
    }
    
    public CompletableFuture<ServiceState> start() {
        return ioToCompletableFuture(service.start())
                .thenApply(r -> new ServiceState((io.jobial.scase.core.ServiceState<IO>) r));
    }
}
